### DISCUSSION TITLE
ilogtail 占用cpu过高

### DISCUSSION THREAD

**Discussion by bad-youngster** [alibaba/ilogtail#558] [Help] [answered]
![image](https://user-images.githubusercontent.com/19181456/206386843-7399e6ca-bc61-4f14-b693-42a058cfbbbd.png)

cpu 8c 
ilogtail 限制 0.6 
yaml配置文件 17个，采用 grok和processor_split_key_value 两种方式采集

**Comment by linrunqi08** [comment]
采集配置可以贴一下吗，另外CPU可以持续监控然后截图。 采集的数据量大概有多少呢

**Comment by bad-youngster** [comment]
![image](https://user-images.githubusercontent.com/19181456/206389307-3afcbf6d-a9ed-41d3-b95c-682675558862.png)
![image](https://user-images.githubusercontent.com/19181456/206389389-d61c0b06-a120-48f9-b506-a872b543d124.png)

**Comment by linrunqi08** [comment]
数据量级是多少呢？grok的配置麻烦也帖一下吧

**Comment by bad-youngster** [comment]
nginx日志实时日志
![image](https://user-images.githubusercontent.com/19181456/206391555-6b31eb29-b4ab-4fe2-8145-21ee1e18110d.png)
最大的34G，其他很小，可以忽略

**Comment by linrunqi08** [comment]
34G是指什么？我想问的是文件写入速率是多少？比如10M/min或者100M/min

**Comment by bad-youngster** [comment]
![image](https://user-images.githubusercontent.com/19181456/206392918-acbe1925-bc89-48fe-b294-6f81a30d98e4.png)

数据量最大的两个文件的写入速度

**Comment by linrunqi08** [comment]
可以发一些样例数据+文字版的样例配置给我们吗，我们这边跑一下看看

**Reply by bad-youngster** [reply]
我是上传到sls，这个不方便给你，其他的数据可以的

**Comment by bad-youngster** [comment]
就是yaml配置和nginx日志文件？

**Reply by linrunqi08** [reply]
是的

**Reply by bad-youngster** [reply]
日志里有敏感信息
配置的yaml可以给你

**Reply by bad-youngster** [reply]
enable: true
inputs:
  - Type: file_log
    LogPath: /var/log/nginx/
    FilePattern:  access.log

processors:
  - Type: processor_split_key_value
    SourceKey: content
    KeepSource: false
    Delimiter: " "
    Separator: ":"


flushers:
  - Type: flusher_sls
    Endpoint: cn-shanghai-intranet.log.aliyuncs.com
    ProjectName: 
    LogstoreName:
~

**Reply by linrunqi08** [reply]
敏感信息可以全量替换一下？整个文件不方便的话，给一段也可以的

**Reply by bad-youngster** [reply]
enable: true
inputs:
  - Type: file_log
    LogPath: /var/log/nginx/
    FilePattern: access.log

processors:
  - Type: processor_grok
    SourceKey: content
    KeepSource: false
    CustomPatterns:
      NGINX_ACCESS: '\"%{DATA:server_addr}\" - %{IPORHOST:remote_addr} - %{USERNAME:remote_user} \[%{HTTPDATE:time_local}\] \"%{DATA:request}\" \"%{DATA:request_method}\" \"%{DATA:request_uri}\" \"%{DATA:server_protocol}\" %{INT:status} %{NUMBER:body_bytes_sent} \"%{DATA:http_referer}\" \"%{DATA:scheme}\" \"%{DATA:ssl_protocol}\" \"%{DATA:ssl_cipher}\" \"%{DATA:http_user_agent}\" \"%{DATA:http_x_forwarded_for}\" %{NUMBER:request_time} %{INT:request_length} %{HOSTPORT:upstream_addr} \[%{NUMBER:upstream_response_time}\] \"%{DATA:proxy_add_x_forwarded_for}\"'
    Match:
      - '%{NGINX_ACCESS}'
    IgnoreParseFailure: false


flushers:
  - Type: flusher_sls
    Endpoint: cn-shanghai-intranet.log.aliyuncs.com
    ProjectName: 
    LogstoreName:

**Reply by bad-youngster** [reply]
server_addr:"172.21.xx.xx" remote_addr:100.116.xxx.xx remote_user:- time_local:[08/Dec/2022:16:21:31 +0800] request:"POST /xxxxx/self/xxxxx HTTP/1.1" status:200 body_bytes_sent:16511 http_referer:"-" scheme:"http" ssl_protocol:"-" ssl_cipher:"-" http_user_agent:"User Agent" http_x_forwarded_for:"59.59.xxx.xx" request_time:0.050 request_time:895 upstream_addr:172.21.xxx.xx:9181 upstream_addr:0.049 proxy_add_x_forwarded_for:"59.59.xxx.xx, 100.116.xxx.xx" app_version:"1.1.44.1" client-type:"LT" platform:"win32" device:"Microsoft Windows 8\xB0\xE6\xB1\xBE\xA3\xBA6.2(9200)" randomKey:"aa11x9xxxxx181xxxxx4xxx1" authorization:"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" sourceFrom:"PC" contentModify:"-"

server_addr:"172.21.xx.xx" remote_addr:100.127.xxx.xx remote_user:- time_local:[08/Dec/2022:16:21:31 +0800] request:"POST /xxxxx/self/groupListWithNumber HTTP/1.1" status:200 body_bytes_sent:125 http_referer:"-" scheme:"http" ssl_protocol:"-" ssl_cipher:"-" http_user_agent:"User Agent" http_x_forwarded_for:"117.90.xxx.xx" request_time:0.012 request_time:851 upstream_addr:172.21.5.9:9181 upstream_addr:0.011 proxy_add_x_forwarded_for:"117.90.xxx.xx, 100.127.xxx.xx" app_version:"1.1.5.1" client-type:"LT" platform:"win32" device:"Microsoft Windows 7\xB0\xE6\xB1\xBE\xA3\xBA6.1(7601)" randomKey:"aa11x9xxxxx181xxxxx4xxx1" authorization:"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" sourceFrom:"PC" contentModify:"-"

server_addr:"172.21.xx.xx" remote_addr:100.116.xxx.xx remote_user:- time_local:[08/Dec/2022:16:21:31 +0800] request:"POST /xxxxx/self/xxxxx HTTP/1.1" status:200 body_bytes_sent:7970 http_referer:"-" scheme:"http" ssl_protocol:"-" ssl_cipher:"-" http_user_agent:"User Agent" http_x_forwarded_for:"117.178.xxx.xx" request_time:0.039 request_time:905 upstream_addr:172.21.xxx.xx:9181 upstream_addr:0.038 proxy_add_x_forwarded_for:"117.178.xxx.xx, 100.116.xxx.xx" app_version:"1.1.44.1" client-type:"LT" platform:"win32" device:"Microsoft Windows 7\xB0\xE6\xB1\xBE\xA3\xBA6.1(7601)" randomKey:"aa11x9xxxxx181xxxxx4xxx1" authorization:"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" sourceFrom:"PC" contentModify:"-"

server_addr:"172.21.xx.xx" remote_addr:100.116.xxx.xx remote_user:- time_local:[08/Dec/2022:16:21:31 +0800] request:"POST /xxxxx/self/xxxxx HTTP/1.1" status:200 body_bytes_sent:2545 http_referer:"-" scheme:"http" ssl_protocol:"-" ssl_cipher:"-" http_user_agent:"User Agent" http_x_forwarded_for:"220.184.xxx.xx" request_time:0.035 request_time:907 upstream_addr:172.21.xxx.xx:9181 upstream_addr:0.034 proxy_add_x_forwarded_for:"220.184.xxx.xx, 100.116.xxx.xx" app_version:"1.1.44.1" client-type:"LT" platform:"win32" device:"Microsoft Windows 8\xB0\xE6\xB1\xBE\xA3\xBA6.2(9200)" randomKey:"aa11x9xxxxx181xxxxx4xxx1" authorization:"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" sourceFrom:"PC" contentModify:"-"

**Comment by bad-youngster** [comment]
使用processor_grok插件，目前来看cpu利用率比较低
![1670492037926_C175BCC5-8687-463c-8945-FA8D510D9A6C](https://user-images.githubusercontent.com/19181456/206413395-6082fb08-b754-4eec-bde7-c5fb05701aa5.png)

使用processor_split_key_value插件，一个yaml配置文件就会导致cpu上升数倍，多个yaml配置文件可能会出现几何倍的上升
单个yaml使用processor_split_key_value解析文件时的cpu：

![1670492069502_A9EECA96-093D-4c52-9E19-E2C32BEA39EB](https://user-images.githubusercontent.com/19181456/206413930-892b4c5f-a813-4203-93db-e7f4a5dfe1c8.png)

**Reply by bad-youngster** [reply]
![image](https://user-images.githubusercontent.com/19181456/206414336-913ad3ab-0a73-46bc-addf-eabfad2db678.png)
这个图，上面图贴错了

**Comment by pj1987111** [comment, marked as answer]
processor_split_key_value插件使用的是string.Split
![image](https://user-images.githubusercontent.com/6268414/229333715-206e8ce8-e77d-4467-b459-2d693cb4f3f1.png)
会造成大量小字符串的分配与释放。改用slice + 偏移可以缓解内存使用，经测试(本机)，10000条下处理速度有一定提升
16.44是split，13.38是slice+偏移
![image](https://user-images.githubusercontent.com/6268414/229333778-f051fd18-4789-4905-ab7b-783666f43881.png)

**Reply by pj1987111** [reply]
我可以来优化一下
